src/plugins/dashboard/public/dashboard_app/top_nav/share/show_share_modal.tsx

```diff
@@ -127,40 +127,51 @@ export function ShowShareModal({
 
   let unsavedStateForLocator: DashboardLocatorParams = {};
 
-  const { dashboardState: unsavedDashboardState, panels: unsavedPanelStates } =
+  const { dashboardState: unsavedDashboardState, panels: panelModifications } =
     dashboardBackup.getState(savedObjectId) ?? {};
 
-  const latestPanels = getDashboardState().panels;
-  const unsavedPanelsMap = unsavedPanelStates
-    ? Object.entries(unsavedPanelStates).reduce((acc, [panelId, unsavedPanel]) => {
-        if (unsavedPanel && latestPanels?.[panelId]) {
-          acc[panelId] = {
-            ...latestPanels[panelId],
-            explicitInput: {
-              ...latestPanels?.[panelId].explicitInput,
-              ...unsavedPanel,
-              id: panelId,
-            },
-          };
-        }
-        return acc;
-      }, {} as DashboardPanelMap)
-    : {};
-
-  const allPanels: DashboardPanelMap = {
-    ...(unsavedDashboardState?.panels ?? {}),
-    ...unsavedPanelsMap,
-  };
+  const allUnsavedPanels = (() => {
+    if (
+      Object.keys(unsavedDashboardState?.panels ?? {}).length === 0 &&
+      Object.keys(panelModifications ?? {}).length === 0
+    ) {
+      // if this dashboard has no modifications or unsaved panels return early. No overrides needed.
+      return;
+    }
+
+    const latestPanels = getDashboardState().panels;
+    // apply modifications to panels.
+    const modifiedPanels = panelModifications
+      ? Object.entries(panelModifications).reduce((acc, [panelId, unsavedPanel]) => {
+          if (unsavedPanel && latestPanels?.[panelId]) {
+            acc[panelId] = {
+              ...latestPanels[panelId],
+              explicitInput: {
+                ...latestPanels?.[panelId].explicitInput,
+                ...unsavedPanel,
+                id: panelId,
+              },
+            };
+          }
+          return acc;
+        }, {} as DashboardPanelMap)
+      : {};
+
+    // this dashboard has unsaved panels.
+    const allUnsavedPanelsMap = {
+      ...latestPanels,
+      ...(unsavedDashboardState?.panels ?? {}),
```

Comment:
Why are you spreading `unsavedDashboardState?.panels`?  How would `unsavedDashboardState?.panels` be different from `panelModifications`?

Reply:
This is pretty messy right now. Likely we can clean this up majorly once the legacy embeddable system is removed.
- `latestPanels` is the latest panels directly from `input`. 
- `unsavedDashboardState?.panels` is the full panels listing if there are _any_ unsaved changes to panels. This is from the legacy system.
- `panelModifications` is a map of all react embeddables with unsaved changes, where each key has been modified.

We can determine if there are any unsaved panels at all by checking if `panelModifications` is populated **or** if `unsavedDashboardState.panels` is populated.

Then if either is populated we return the complete panels. When spreading, we can _probably_ omit `unsavedDashboardState?.panels` - not because it's covered in `panelModifications`, but because it would be covered by `latestPanels`.

Let me test this.

In the future, when the old system is removed we will serialize `panelModifications` here directly because that will contain everything. Then on load time we can zip those with the panels loaded from the saved object rather than replacing them wholesale.

Reply:
Update, yes I could technically remove this, just had to update a test to better match the actual behaviour.